pkg/core/consensus/reduction/reduction.go

```diff
@@ -81,6 +85,17 @@ func (r *Reduction) verifyWithDelay(candidate *block.Block, step uint8) ([]byte,
 		return nil, errEmptyBlockHash
 	}
 
+	hash, err := candidate.CalculateHash()
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if we have already verified this block.
+	if len(r.VerifiedHash) == 32 && !bytes.Equal(r.VerifiedHash, block.EmptyHash[:]) && bytes.Equal(r.VerifiedHash, hash) {
```

Comment:
I agree with this. But what if previous step verified a hash that differ from the current step hash? Should the error be raised immediately instead of call verifyFn again?

Reply:
In theory, this is an edge case that is allowed to happen but I'll add a warning for now.